File: mobile/android/app/src/main/kotlin/com/masim05/bloodpressure/mobile/core/flow/AppFlows.kt
package com.masim05.bloodpressure.mobile.core.flow

import com.masim05.bloodpressure.mobile.core.model.ApiError
import com.masim05.bloodpressure.mobile.core.model.AuthMode
import com.masim05.bloodpressure.mobile.core.model.AppResult
import com.masim05.bloodpressure.mobile.core.model.HistoryFilter
import com.masim05.bloodpressure.mobile.core.model.Measurement
import com.masim05.bloodpressure.mobile.core.model.MeasurementDetail
import com.masim05.bloodpressure.mobile.core.model.MeasurementImage
import com.masim05.bloodpressure.mobile.core.model.Session
import com.masim05.bloodpressure.mobile.core.ports.AuthGateway
import com.masim05.bloodpressure.mobile.core.ports.CameraGateway
import com.masim05.bloodpressure.mobile.core.ports.HistoryGateway
import com.masim05.bloodpressure.mobile.core.ports.MeasurementDetailGateway
import com.masim05.bloodpressure.mobile.core.ports.MeasurementUploadGateway
import com.masim05.bloodpressure.mobile.core.ports.SessionStore
import com.masim05.bloodpressure.mobile.core.validation.ValidationError
import com.masim05.bloodpressure.mobile.core.validation.ValidationResult
import com.masim05.bloodpressure.mobile.core.validation.Validators

enum class Route {
    Auth,
    Guide,
    Camera,
    History,
    MeasurementDetail,
}

data class ScreenState(
    val route: Route,
    val authMode: AuthMode = AuthMode.Login,
    val session: Session? = null,
    val error: ApiError? = null,
    val validationError: ValidationError? = null,
    val measurements: List<Measurement> = emptyList(),
    val measurementDetail: MeasurementDetail? = null,
    val filter: HistoryFilter = HistoryFilter(),
    val lastUploadId: String? = null,
)

class AuthFlow(
    private val authGateway: AuthGateway,
    private val sessionStore: SessionStore,
) {
    fun signIn(email: String, password: String): ScreenState = authenticate(
        email = email,
        password = password,
        routeOnSuccess = Route.Guide,
        authMode = AuthMode.NewAccount,
    ) { authGateway.signIn(email.trim(), password) }

    fun logIn(email: String, password: String): ScreenState = authenticate(
        email = email,
        password = password,
        routeOnSuccess = Route.Camera,
        authMode = AuthMode.Login,
    ) { authGateway.logIn(email.trim(), password) }

    private fun authenticate(
        email: String,
        password: String,
        routeOnSuccess: Route,
        authMode: AuthMode,
        request: () -> AppResult<Session>,
    ): ScreenState {
        Validators.email(email).let { if (it is ValidationResult.Invalid) return ScreenState(Route.Auth, authMode = authMode, validationError = it.reason) }
        Validators.password(password).let { if (it is ValidationResult.Invalid) return ScreenState(Route.Auth, authMode = authMode, validationError = it.reason) }
        return when (val result = request()) {
            is AppResult.Success -> {
                sessionStore.save(result.value)
                ScreenState(routeOnSuccess, session = result.value)
            }
            is AppResult.Failure -> ScreenState(Route.Auth, authMode = authMode, error = result.error)
        }
    }
}

class GuideFlow(private val sessionStore: SessionStore) {
    fun enterGuide(): ScreenState = sessionStore.load()?.let { ScreenState(Route.Guide, session = it) } ?: ScreenState(Route.Auth)
    fun continueToCamera(): ScreenState = sessionStore.load()?.let { ScreenState(Route.Camera, session = it) } ?: ScreenState(Route.Auth)
}

class CaptureFlow(
    private val sessionStore: SessionStore,
    private val cameraGateway: CameraGateway,
    private val uploadGateway: MeasurementUploadGateway,
) {
    fun enterCamera(): ScreenState = sessionStore.load()?.let { ScreenState(Route.Camera, session = it) } ?: ScreenState(Route.Auth)

    fun history(): ScreenState = sessionStore.load()?.let { ScreenState(Route.History, session = it) } ?: ScreenState(Route.Auth)

    fun captureAndUpload(): ScreenState {
        val session = sessionStore.load() ?: return ScreenState(Route.Auth)
        if (!cameraGateway.isReady()) {
            return ScreenState(
                route = Route.Camera,
                session = session,
                error = ApiError(
                    code = "camera_not_ready",
                    message = "Camera is not ready for capture.",
                    source = com.masim05.bloodpressure.mobile.core.model.ApiErrorSource.Unexpected,
                ),
            )
        }
        return when (val camera = cameraGateway.openCamera()) {
            is AppResult.Failure -> ScreenState(Route.Camera, session = session, error = camera.error)
            is AppResult.Success -> upload(session, camera.value)
        }
    }

    private fun upload(session: Session, image: MeasurementImage): ScreenState {
        Validators.image(image).let { if (it is ValidationResult.Invalid) return ScreenState(Route.Camera, session = session, validationError = it.reason) }
        return when (val upload = uploadGateway.upload(session, image)) {
            is AppResult.Success -> ScreenState(Route.History, session = session, lastUploadId = upload.value)
            is AppResult.Failure -> ScreenState(Route.Camera, session = session, error = upload.error)
        }
    }
}

class HistoryFlow(
    private val sessionStore: SessionStore,
    private val historyGateway: HistoryGateway,
) {
    fun load(filter: HistoryFilter): ScreenState {
        val session = sessionStore.load() ?: return ScreenState(Route.Auth)
        Validators.historyFilter(filter).let { if (it is ValidationResult.Invalid) return ScreenState(Route.History, session = session, validationError = it.reason, filter = filter) }
        return when (val history = historyGateway.list(session, filter)) {
            is AppResult.Success -> ScreenState(Route.History, session = session, measurements = history.value, filter = filter)
            is AppResult.Failure -> ScreenState(Route.History, session = session, error = history.error, filter = filter)
        }
    }

    fun rowOpensDetail(measurement: Measurement): ScreenState = sessionStore.load()?.let {
        ScreenState(Route.MeasurementDetail, session = it, measurementDetail = measurement.toDetail())
    } ?: ScreenState(Route.Auth)

    private fun Measurement.toDetail(): MeasurementDetail = MeasurementDetail(
        id = id,
        status = status,
        systolic = systolic,
        diastolic = diastolic,
        pulse = pulse,
        armSide = armSide,
        measurementTime = measurementTime,
        savedAt = savedAt,
        imageUrl = "",
        recognitionError = null,
    )
}

class MeasurementDetailFlow(
    private val sessionStore: SessionStore,
    private val detailGateway: MeasurementDetailGateway,
) {
    fun load(measurementId: String): ScreenState {
        val session = sessionStore.load() ?: return ScreenState(Route.Auth)
        return when (val detail = detailGateway.get(session, measurementId)) {
            is AppResult.Success -> ScreenState(Route.MeasurementDetail, session = session, measurementDetail = detail.value)
            is AppResult.Failure -> ScreenState(Route.MeasurementDetail, session = session, error = detail.error)
        }
    }

    fun save(detail: MeasurementDetail): ScreenState {
        val session = sessionStore.load() ?: return ScreenState(Route.Auth)
        return when (val saved = detailGateway.save(session, detail)) {
            is AppResult.Success -> ScreenState(Route.MeasurementDetail, session = session, measurementDetail = saved.value)
            is AppResult.Failure -> ScreenState(Route.MeasurementDetail, session = session, measurementDetail = detail, error = saved.error)
        }
    }
}

File: mobile/ios/Blood pressure tracker/Blood pressure tracker/AppFlows.swift
// AppFlows.swift
// Blood pressure tracker
// Business-logic flows mirroring Android's AppFlows.kt

import Foundation

// MARK: - Auth flow

class AuthFlow {
    private let authGateway: AuthGateway
    private let sessionStore: SessionStore

    init(authGateway: AuthGateway, sessionStore: SessionStore) {
        self.authGateway = authGateway
        self.sessionStore = sessionStore
    }

    func signIn(email: String, password: String) -> ScreenState {
        authenticate(email: email, password: password, routeOnSuccess: .guide, authMode: .newAccount) {
            self.authGateway.signIn(email: email.trimmingCharacters(in: .whitespaces), password: password)
        }
    }

    func logIn(email: String, password: String) -> ScreenState {
        authenticate(email: email, password: password, routeOnSuccess: .camera, authMode: .login) {
            self.authGateway.logIn(email: email.trimmingCharacters(in: .whitespaces), password: password)
        }
    }

    private func authenticate(
        email: String,
        password: String,
        routeOnSuccess: Route,
        authMode: AuthMode,
        request: () -> AppResult<Session>
    ) -> ScreenState {
        if case .invalid(let reason) = Validators.email(email) {
            return ScreenState(route: .auth, authMode: authMode, validationError: reason)
        }
        if case .invalid(let reason) = Validators.password(password) {
            return ScreenState(route: .auth, authMode: authMode, validationError: reason)
        }
        switch request() {
        case .success(let session):
            sessionStore.save(session)
            return ScreenState(route: routeOnSuccess, session: session)
        case .failure(let error):
            return ScreenState(route: .auth, authMode: authMode, error: error)
        }
    }
}

// MARK: - Guide flow

class GuideFlow {
    private let sessionStore: SessionStore

    init(sessionStore: SessionStore) {
        self.sessionStore = sessionStore
    }

    func enterGuide() -> ScreenState {
        guard let session = sessionStore.load() else { return ScreenState(route: .auth) }
        return ScreenState(route: .guide, session: session)
    }

    func continueToCamera() -> ScreenState {
        guard let session = sessionStore.load() else { return ScreenState(route: .auth) }
        return ScreenState(route: .camera, session: session)
    }
}

// MARK: - Capture flow

class CaptureFlow {
    private let sessionStore: SessionStore
    private let cameraGateway: CameraGateway
    private let uploadGateway: MeasurementUploadGateway

    init(sessionStore: SessionStore, cameraGateway: CameraGateway, uploadGateway: MeasurementUploadGateway) {
        self.sessionStore = sessionStore
        self.cameraGateway = cameraGateway
        self.uploadGateway = uploadGateway
    }

    func enterCamera() -> ScreenState {
        guard let session = sessionStore.load() else { return ScreenState(route: .auth) }
        return ScreenState(route: .camera, session: session)
    }

    func history() -> ScreenState {
        guard let session = sessionStore.load() else { return ScreenState(route: .auth) }
        return ScreenState(route: .history, session: session)
    }

    func captureAndUpload() -> ScreenState {
        guard let session = sessionStore.load() else { return ScreenState(route: .auth) }
        if !cameraGateway.isReady() {
            return ScreenState(
                route: .camera,
                session: session,
                error: ApiError(code: "camera_not_ready", message: "camera_not_ready", source: .unexpected)
            )
        }
        switch cameraGateway.openCamera() {
        case .failure(let error):
            return ScreenState(route: .camera, session: session, error: error)
        case .success(let image):
            return upload(session: session, image: image)
        }
    }

    private func upload(session: Session, image: MeasurementImage) -> ScreenState {
        if case .invalid(let reason) = Validators.image(image) {
            return ScreenState(route: .camera, session: session, validationError: reason)
        }
        switch uploadGateway.upload(session: session, image: image) {
        case .success(let id):
            return ScreenState(route: .history, session: session, lastUploadId: id)
        case .failure(let error):
            return ScreenState(route: .camera, session: session, error: error)
        }
    }
}

// MARK: - History flow

class HistoryFlow {
    private let sessionStore: SessionStore
    private let historyGateway: HistoryGateway

    init(sessionStore: SessionStore, historyGateway: HistoryGateway) {
        self.sessionStore = sessionStore
        self.historyGateway = historyGateway
    }

    func load(filter: HistoryFilter) -> ScreenState {
        guard let session = sessionStore.load() else { return ScreenState(route: .auth) }
        if case .invalid(let reason) = Validators.historyFilter(filter) {
            return ScreenState(route: .history, session: session, validationError: reason, filter: filter)
        }
        switch historyGateway.list(session: session, filter: filter) {
        case .success(let measurements):
            return ScreenState(route: .history, session: session, measurements: measurements, filter: filter)
        case .failure(let error):
            return ScreenState(route: .history, session: session, error: error, filter: filter)
        }
    }

    func rowOpensDetail(_ measurement: Measurement) -> ScreenState {
        guard let session = sessionStore.load() else { return ScreenState(route: .auth) }
        let detail = MeasurementDetail(
            id: measurement.id,
            status: measurement.status,
            systolic: measurement.systolic,
            diastolic: measurement.diastolic,
            pulse: measurement.pulse,
            armSide: measurement.armSide,
            measurementTime: measurement.measurementTime,
            savedAt: measurement.savedAt,
            imageUrl: "",
            recognitionError: nil
        )
        return ScreenState(route: .measurementDetail, session: session, measurementDetail: detail)
    }
}

// MARK: - Measurement detail flow

class MeasurementDetailFlow {
    private let sessionStore: SessionStore
    private let detailGateway: MeasurementDetailGateway

    init(sessionStore: SessionStore, detailGateway: MeasurementDetailGateway) {
        self.sessionStore = sessionStore
        self.detailGateway = detailGateway
    }

    func load(measurementId: String) -> ScreenState {
        guard let session = sessionStore.load() else { return ScreenState(route: .auth) }
        switch detailGateway.get(session: session, measurementId: measurementId) {
        case .success(let detail):
            return ScreenState(route: .measurementDetail, session: session, measurementDetail: detail)
        case .failure(let error):
            return ScreenState(route: .measurementDetail, session: session, error: error)
        }
    }

    func save(detail: MeasurementDetail) -> ScreenState {
        guard let session = sessionStore.load() else { return ScreenState(route: .auth) }
        switch detailGateway.save(session: session, detail: detail) {
        case .success(let saved):
            return ScreenState(route: .measurementDetail, session: session, measurementDetail: saved)
        case .failure(let error):
            return ScreenState(route: .measurementDetail, session: session, error: error, measurementDetail: detail)
        }
    }
}
